File: Golang/decorator/decorator.go
package decorator

type CondimentDecorator interface {
	Beverage
	AddOnDescription() string
}

type Milk struct {
	beverage Beverage
}

func NewMilk(beverage Beverage) Beverage {
	return &Milk{beverage: beverage}
}

func (m *Milk) Cost() float64 {
	return 0.1 + m.beverage.Cost()
}

func (m *Milk) GetDescription() string {
	return m.beverage.GetDescription() + m.AddOnDescription()
}

func (m *Milk) AddOnDescription() string {
	return "\n + AddOn Milk"
}

type Mocha struct {
	beverage Beverage
}

func NewMocha(beverage Beverage) Beverage {
	return &Mocha{beverage: beverage}
}

func (m *Mocha) Cost() float64 {
	return 0.2 + m.beverage.Cost()
}

func (m *Mocha) GetDescription() string {
	return m.beverage.GetDescription() + m.AddOnDescription()
}

func (m *Mocha) AddOnDescription() string {
	return "\n + AddOn Mocha"
}

type Whip struct {
	beverage Beverage
}

func NewWhip(beverage Beverage) Beverage {
	return &Whip{beverage: beverage}
}

func (m *Whip) Cost() float64 {
	return 0.2 + m.beverage.Cost()
}

func (m *Whip) GetDescription() string {
	return m.beverage.GetDescription() + m.AddOnDescription()
}

func (m *Whip) AddOnDescription() string {
	return "\n + AddOn Whip"
}

File: Python/decorator/decorator.py
from beverage import Beverage
from condiment_decorator import CondimentDecorator


class Milk(CondimentDecorator):
    def __init__(self, beverage: Beverage):
        self.beverage = beverage

    def getDescription(self) -> str:
        return self.beverage.getDescription() + ', Milk'

    def cost(self) -> float:
        return 0.10 + self.beverage.cost()


class Mocha(CondimentDecorator):
    def __init__(self, beverage: Beverage):
        self.beverage = beverage

    def getDescription(self) -> str:
        return self.beverage.getDescription() + ', Mocha'

    def cost(self) -> float:
        return 0.20 + self.beverage.cost()


class Soy(CondimentDecorator):
    def __init__(self, beverage: Beverage):
        self.beverage = beverage

    def getDescription(self) -> str:
        return self.beverage.getDescription() + ', Soy'

    def cost(self) -> float:
        return 0.15 + self.beverage.cost()


class Whip(CondimentDecorator):
    def __init__(self, beverage: Beverage):
        self.beverage = beverage

    def getDescription(self) -> str:
        return self.beverage.getDescription() + ', Whip'

    def cost(self) -> float:
        return 0.10 + self.beverage.cost()
